rmw_connext_cpp/src/rmw_init.cpp

```diff
@@ -111,24 +116,40 @@ rmw_init_options_fini(rmw_init_options_t * init_options)
 rmw_ret_t
 rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
 {
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(options, RMW_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(options, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    options->implementation_identifier,
+    "expected initialized init options",
+    return RMW_RET_INVALID_ARGUMENT);
   RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
     options,
     options->implementation_identifier,
     rti_connext_identifier,
     return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  if (NULL != context->implementation_identifier) {
+    RMW_SET_ERROR_MSG("expected a zero-initialized context");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  auto cleanup = rcpputils::make_scope_exit(
+    [context]() {
+      delete context->impl;
+      *context = rmw_get_zero_initialized_context();
+    });
```

Comment:
can you move this after the line where the context impl is created?
or add a check before this, failing if `context->impl` wasn't a `nullptr`.

Reply:
So, `delete nullptr` is a no-op and this object is created at a point where we've assumed `context` is zero-initialized. As I said elsewhere, we can split this into two `scope_exit` instances, but your concern about `context->impl` not being `nullptr` is orthogonal. I'm not entirely convinced we have to deal with bogus code that breaks API preconditions. We can `assert` though.

Reply:
> We can assert though.

assert sounds fine.

>  I'm not entirely convinced we have to deal with bogus code that breaks API preconditions

If the extra check needed to avoid UB is only one line, I prefer adding it.

Reply:
Fair enough, see https://github.com/ros2/rmw_connext/pull/432/commits/c1eb006f475586bd64392eb5f85a469d40648762